src/nacl/bindings/crypto_generichash.py

```diff
@@ -122,6 +122,26 @@ def generichash_blake2b_salt_personal(data,
     return ffi.buffer(digest, digest_size)[:]
 
 
+class Blake2State(object):
+    """
+    Python-level wrapper for the crypto_generichash_blake2b state buffer
+    """
+    __slots__ = ['_statebuf', 'digest_size']
+
+    def __init__(self, digest_size):
+        self._statebuf = ffi.new("unsigned char[]",
+                                 crypto_generichash_STATEBYTES)
+        self.digest_size = digest_size
+
+    def __copy__(self):
```

Comment:
Since this is copying the state to a new object rather than a reference isn't this more correctly `__deepcopy__`?

Reply:
The correct copy behaviour is for the copy of the state to be a different object, like happens for lists. copy.deepcopy already works correctly.

Reply:
The semantics of `__copy__` are normally that you get a new object that contains identical references for members. `__deepcopy__` is a new object and all of its members are also new objects. So this seems like a deepcopy to me?

Reply:
IMO this is just like what happens for mutable types, since in their case copy.copy creates a new instance, with the same initial content:
```
$ python3
Python 3.7.2+ (default, Feb  2 2019, 14:31:48)
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> l1 = [1, 2, 3]
>>> import copy
>>> l2 = copy.copy(l1)
>>> l1 is l2
False
>>> l2.append(4)
>>> l1
[1, 2, 3]
>>> l2
[1, 2, 3, 4]
>>> d1 = {1:1, 2:2}
>>> d2 = copy.copy(d1)
>>> d2 is d1
False
>>> d2[3] = 3
>>> d1
{1: 1, 2: 2}
>>> d2
{1: 1, 2: 2, 3: 3}
>>>
```

Reply:
The reference to a correct behaviour of deepcopy in https://github.com/pyca/pynacl/pull/514#discussion_r262255096 is outdated, since that worked correctly in the aborted bytearray attempt.

Reply:
Yes, copy makes a new top level object, but the documentation states that it uses references for the members of that new object. So `_statebuf` would be the same reference, rather than a new object itself. deepcopy, by contrast, also creates copies of all members recursively.

This is a pedantic argument since the implementation here is correct -- I just think we should be using `__deepcopy__` as the name of our method and calling `deepcopy` the function, which will use `__deepcopy__`.